src/binfhe/lib/rgsw-acc-lmkcdey.cpp

```diff
@@ -0,0 +1,341 @@
+//==================================================================================
+// BSD 2-Clause License
+//
+// Copyright (c) 2014-2022, NJIT, Duality Technologies Inc. and other contributors
+//
+// All rights reserved.
+//
+// Author TPOC: contact@openfhe.org
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this
+//    list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+//    this list of conditions and the following disclaimer in the documentation
+//    and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+#include "rgsw-acc-lmkcdey.h"
+
+#include <string>
+
+namespace lbcrypto {
+
+// Key generation as described in https://eprint.iacr.org/2022/198
+RingGSWACCKey RingGSWAccumulatorLMKCDEY::KeyGenAcc(const std::shared_ptr<RingGSWCryptoParams> params,
+                                              const NativePoly& skNTT, ConstLWEPrivateKey LWEsk) const {
+    auto sv     = LWEsk->GetElement();
+    int32_t mod = sv.GetModulus().ConvertToInt();
+    int32_t modHalf = mod >> 1;
+    uint32_t N = params->GetN();
+    uint32_t n = sv.GetLength();
+
+    // dim2, 0: for RGSW(X^si), 1: for automorphism keys
+    // only w automorphism keys required
+    // allocates (n - w) more memory for pointer (not critical for performance)
+    RingGSWACCKey ek = std::make_shared<RingGSWACCKeyImpl>(1, 2, n);  
+
+#pragma omp parallel for
+    for (size_t i = 0; i < n; ++i) {
+        int32_t s = (int32_t)sv[i].ConvertToInt();
+        if (s > modHalf) {
+            s -= mod;
+        }
+
+        (*ek)[0][0][i] = KeyGenLMKCDEY(params, skNTT, s);
+    }
+    
+    NativeInteger gen = NativeInteger(5);
+    
+    (*ek)[0][1][0] = KeyGenAuto(params, skNTT, 2*N - gen.ConvertToInt());
+
+    // m_window: window size, consider parameterization in the future
+    for (size_t i = 1; i < m_window+1; i++)
```

Comment:
Why not use OMP parallelization here?

Reply:
Of course we can use parallelization here. I'll add it.